File: src/infogrep/infogrep.py
import argparse
import re
import sys
import json
import os

from . import utils

parser = argparse.ArgumentParser(description="Grep for sensitive info")
parser.add_argument('-i', '--input', help="file or directory to scan")
parser.add_argument('-p', '--pattern', default='secrets', help="pick a pattern from .config/infogrep.patterns.json")
parser.add_argument('-a', '--add-pattern', help="add a pattern file to .config/infogrep.patterns.json (provide name:path [-a name:/path/to/pattern.yml])")

args = parser.parse_args()

def add_custom_pattern(name, path):
    config_dir = os.path.expanduser("~/.config")
    config_file = os.path.join(config_dir, "infogrep.patterns.json")

    # Ensure the config directory exists
    os.makedirs(config_dir, exist_ok=True)

    # Load existing patterns or create an empty dict if the file doesn't exist
    if os.path.exists(config_file):
        with open(config_file, 'r') as f:
            patterns = json.load(f)
    else:
        patterns = {}

    # Add or update the new pattern
    patterns[name] = path

    # Save the updated patterns back to the file
    with open(config_file, 'w') as f:
        json.dump(patterns, f, indent=2)

    print(f"Pattern '{name}' added successfully.")

def greppin(content: str, pattern: dict) -> list:
    matches = re.finditer(pattern['pattern']['regex'], content, re.MULTILINE)
    return [(match.group(), match.start()) for match in matches]  # Return match and its position

def truncate_match(match: str, max_chars: int = 300) -> str:
    if len(match) > max_chars:
        truncated = match[:max_chars]
        return f"{truncated}... (truncated, {len(match) - max_chars} more characters)"
    return match

PatternName = args.pattern

def main():
    if args.add_pattern:
        try:
            name, path = args.add_pattern.split(':')
            add_custom_pattern(name.strip(), path.strip())
            return
        except ValueError:
            print("Error: Invalid format for add-pattern. Use: name:/path/to/pattern.yml")
            return

    print(utils.logo)
    Patterns = utils.getPatterns(PatternName)

    if args.input:
        files2Scan = utils.get_all_abs_paths(args.input)
        for path in files2Scan:
            print(f"\rScanning [ {path} ]", end="", flush=True)
            try:
                with open(path, 'r') as file:
                    content = file.read()
                for pattern in Patterns['patterns']:
                    results = greppin(content, pattern)
                    if results:
                        print()  # Move to the next line before printing results
                        for res, pos in results:
                            truncated_res = truncate_match(res)
                            utils.print_result(pattern, truncated_res, path, pos)
            except Exception as e:
                print(f"\nAn error occurred while processing {path}: {e}")
    else:
        # Read from stdin if no input file/directory is specified
        content = sys.stdin.read()
        for pattern in Patterns['patterns']:
            results = greppin(content, pattern)
            for res, pos in results:
                truncated_res = truncate_match(res)
                utils.print_result(pattern, truncated_res, "stdin", pos)

    print()  # Print a newline at the end to move the cursor to the next line

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print('\rAborting ...')
        sys.exit(0)

File: cmd/infogrep.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strings"

	"github.com/Giardi77/infogrep/pkg/utils"
)

var (
	inputFlag      string
	patternFlag    string
	addPatternFlag string
	truncateFlag   int
)

func init() {
	flag.StringVar(&inputFlag, "i", "", "file or directory to scan")
	flag.StringVar(&inputFlag, "input", "", "file or directory to scan")
	flag.StringVar(&patternFlag, "p", "secrets", "pick a pattern from .config/infogrep.patterns.json")
	flag.StringVar(&patternFlag, "pattern", "secrets", "pick a pattern from .config/infogrep.patterns.json")
	flag.StringVar(&addPatternFlag, "a", "", "add a pattern file to .config/infogrep.patterns.json (provide name:path)")
	flag.StringVar(&addPatternFlag, "add-pattern", "", "add a pattern file to .config/infogrep.patterns.json (provide name:path)")
	flag.IntVar(&truncateFlag, "t", 400, "Truncation length for output (0 for no truncation)")
	flag.IntVar(&truncateFlag, "truncate", 400, "Truncation length for output (0 for no truncation)")
}

func greppin(content string, pattern utils.Pattern) [][]int {
	re := regexp.MustCompile(pattern.Regex)
	return re.FindAllStringIndex(content, -1)
}

func truncateMatch(match string, maxChars int) string {
	if maxChars == 0 || len(match) <= maxChars {
		return match
	}
	return fmt.Sprintf("%s... (truncated, %d more characters)", match[:maxChars], len(match)-maxChars)
}

func main() {
	flag.Parse()

	if addPatternFlag != "" {
		parts := strings.SplitN(addPatternFlag, ":", 2)
		if len(parts) != 2 {
			fmt.Println("Error: Invalid format for add-pattern. Use: name:/path/to/pattern.yml")
			return
		}
		err := utils.AddCustomPattern(parts[0], parts[1])
		if err != nil {
			fmt.Printf("Error adding custom pattern: %v\n", err)
		}
		return
	}

	fmt.Println(utils.Logo)

	patterns, err := utils.GetPatterns(patternFlag)
	if err != nil {
		fmt.Printf("Error loading patterns: %v\n", err)
		return
	}

	if inputFlag != "" {
		files, err := utils.GetAllAbsPaths(inputFlag)
		if err != nil {
			fmt.Printf("Error getting file paths: %v\n", err)
			return
		}
		for _, path := range files {
			fmt.Printf("\rScanning [ %s ]", path)
			content, err := ioutil.ReadFile(path)
			if err != nil {
				fmt.Printf("\nError reading file %s: %v\n", path, err)
				continue
			}
			for _, pattern := range patterns.Patterns {
				results := greppin(string(content), pattern)
				if len(results) > 0 {
					fmt.Println() // Move to the next line before printing results
					for _, match := range results {
						result := string(content[match[0]:match[1]])
						truncatedResult := truncateMatch(result, truncateFlag)
						utils.PrintResult(pattern, truncatedResult, path, match[0])
					}
				}
			}
		}
	} else {
		reader := bufio.NewReader(os.Stdin)
		content, _ := ioutil.ReadAll(reader)
		for _, pattern := range patterns.Patterns {
			results := greppin(string(content), pattern)
			for _, match := range results {
				result := string(content[match[0]:match[1]])
				truncatedResult := truncateMatch(result, truncateFlag)
				utils.PrintResult(pattern, truncatedResult, "stdin", match[0])
			}
		}
	}

	fmt.Println() // Print a newline at the end
}
